numpy/core/src/_simd/_simd.dispatch.c.src

```diff
@@ -0,0 +1,367 @@
+ * #bsfx      = b8, b8, b16, b16, b32, b32, b64, b64, b32, b64#
+ * #simd_sup  = 1,  1,  1,   1,   1,   1,   1,   1,   1,   NPY_SIMD_F64#
+ * #sat_sup   = 1,  1,  1,   1,   0,   0,   0,   0,   0,   0#
+ * #mul_sup   = 1,  1,  1,   1,   1,   1,   0,   0,   1,   1#
+ * #div_sup   = 0,  0,  0,   0,   0,   0,   0,   0,   1,   1#
+ * #shl_imm   = 0,  0,  15,  15,  31,  31,  63,  63,  0,   0#
+ * #shr_imm   = 0,  0,  16,  16,  32,  32,  64,  64,  0,   0#
+ */
+#if @simd_sup@
+
+/***************************
+ * Memory
+ ***************************/
+/**begin repeat1
+ * # intrin = load, loada, loads, loadl, store, storea, stores, storel, storeh#
+ */
+SIMD_INTRIN_DEF(@intrin@_@sfx@)
+/**end repeat1**/
+
+/***************************
+ * Misc
+ ***************************/
+/**begin repeat1
+ * #sfx_to     = u8, s8, u16, s16, u32, s32, u64, s64, f32, f64#
+ * #simd_sup2  = 1,  1,  1,   1,   1,   1,   1,   1,   1,   NPY_SIMD_F64#
+ */
+#if @simd_sup2@
+SIMD_INTRIN_DEF(reinterpret_@sfx_to@_@sfx@)
+#endif // simd_sup2
+/**end repeat1**/
+
+/**begin repeat1
+ * # intrin = set, setf, setall, zero, select#
+ */
+SIMD_INTRIN_DEF(@intrin@_@sfx@)
+/**end repeat1**/
+
+/***************************
+ * Reorder
+ ***************************/
+/**begin repeat1
+ * # intrin = combinel, combineh, combine, zip#
+ */
+SIMD_INTRIN_DEF(@intrin@_@sfx@)
+/**end repeat1**/
+
+SIMD_INTRIN_DEF(cvt_@sfx@_@bsfx@)
+SIMD_INTRIN_DEF(cvt_@bsfx@_@sfx@)
+
+/***************************
+ * Operators
+ ***************************/
+#if @shl_imm@ > 0
+/**begin repeat1
+ * # intrin = shl, shr, shli, shri#
+ */
+SIMD_INTRIN_DEF(@intrin@_@sfx@)
+/**end repeat1**/
+#endif // shl_imm
+
+/**begin repeat1
+ * #intrin  = and, or, xor, not, cmpeq, cmpneq, cmpgt, cmpge, cmplt, cmple#
+ */
+SIMD_INTRIN_DEF(@intrin@_@sfx@)
+/**end repeat1**/
+
+/***************************
+ * Conversion
+ ***************************/
+SIMD_INTRIN_DEF(cvt_@sfx@_@bsfx@)
+SIMD_INTRIN_DEF(cvt_@bsfx@_@sfx@)
+
+/***************************
+ * Arithmetic
+ ***************************/
+/**begin repeat1
+ * #intrin  = add, sub#
+ */
+SIMD_INTRIN_DEF(@intrin@_@sfx@)
+/**end repeat1**/
+
+#if @sat_sup@
+/**begin repeat1
+ * #intrin  = adds, subs#
+ */
+SIMD_INTRIN_DEF(@intrin@_@sfx@)
+/**end repeat1**/
+#endif // sat_sup
+
+#if @mul_sup@
+SIMD_INTRIN_DEF(mul_@sfx@)
+#endif // mul_sup
+
+#if @div_sup@
+SIMD_INTRIN_DEF(div_@sfx@)
+#endif // div_sup
+
+#endif // simd_sup
+/**end repeat**/
+
+/***************************
+ * Variant
+ ***************************/
+SIMD_INTRIN_DEF(cleanup)
+/***************************/
+{NULL, NULL, 0, NULL}
+}; // PyMethodDef
+
+#endif // NPY_SIMD
+
+/*************************************************************************
+ * Defining a separate module for each target
+ *************************************************************************/
+NPY_VISIBILITY_HIDDEN void
+NPY_CPU_DISPATCH_CURFX(simd_attach_module)(PyObject *dict)
```

Comment:
Can this be made to return `-1` on error as
```suggestion
NPY_VISIBILITY_HIDDEN int
NPY_CPU_DISPATCH_CURFX(simd_attach_module)(PyObject *dict)
```

Reply:
Yes, sure good suggestion, so we can stop the load rest of the submodules in case if an error occurred.
I guess we can change the runtime call to the following:
```Diff
--NPY_CPU_DISPATCH_CALL_ALL(simd_attach_module, (targets))
++NPY_CPU_DISPATCH_CALL_ALL(int mod_err = simd_attach_module, (targets);, if (mod_err) goto err)
````
or we can make a custom macro on the top of abstract macro `NPY__CPU_DISPATCH_CALL`

Reply:
Is there some way you can rework the macro to be an expression rather than a statement? It looks like right now you define it as:
```
      #define DISPATCH_CALL_ALL_CB(CHECK, TARGET_NAME, FN, ARGS) \
        if (CHECK) { NPY_CAT(NPY_CAT(FN, _), TARGET_NAME) ARGS; }
      #define DISPATCH_CALL_BASELINE_ALL_CB(FN, ARGS) \
        FN NPY_EXPAND(ARGS);
```
but you could use instead:
```
      #define DISPATCH_CALL_ALL_CB(CHECK, TARGET_NAME, FN, ARGS) \
        (CHECK) ? (NPY_CAT(NPY_CAT(FN, _), TARGET_NAME) ARGS) : 
      #define DISPATCH_CALL_BASELINE_ALL_CB(FN, ARGS) \
        (FN NPY_EXPAND(ARGS))
```

Reply:
using expressions make it less hackable.
plus the current suggestion code returns one value from one call, you already fixed it in #17201

Reply:
> using expressions make it less hackable.

That can be argued as a feature. While I agree your version above would work, I think gotos inside macro arguments are perhaps a step too far.

And if we did decide that the code snipped you post above was the only way, we could always define the macros you need to pull it off locally.

Reply:
> I think gotos inside macro arguments are perhaps a step too far.

you wouldn't able to stop the preprocessor iterations of `NPY_CPU_DISPATCH_CALL_ALL` without internal
return or goto and #17201 eliminate this feature.

> we could always define the macros you need to pull it off locally.

I guess that we will have to do after #17201 getting merged, I can just gather the runtime call of
`simd_attach_module` with `CHECK_NON_SUPPORTED` into one callback.